manifold/src/boolean_result.cpp

```diff
@@ -674,10 +679,49 @@ Manifold::Impl Boolean3::Result(Manifold::OpType op) const {
   Timer simplify;
   simplify.Start();
 
-  outR.DuplicateMeshIDs();
-
   outR.SimplifyTopology();
 
+  // Update meshID
+  // we remap them into indices starting from 0. The exact value value is not
+  // important as long as
+  // 1. They are distinct
+  // 2. `originalID[meshID]` is the original mesh ID of the triangle
+  // Index starting from 0 is chosen because I want the kernel part as simple as
+  // possible.
+  VecDH<int> meshIDs;
+  VecDH<int> original;
+  for (auto &entry : inP_.meshRelation_.originalID) {
+    meshIDs.push_back(entry.first | 1 << 30);
+    original.push_back(entry.second);
+  }
+  for (auto &entry : inQ_.meshRelation_.originalID) {
+    meshIDs.push_back(entry.first);
+    original.push_back(entry.second);
+  }
+  // in general, meshIDs should be small
+  thrust::sort_by_key(thrust::host, meshIDs.begin(), meshIDs.end(), original.begin());
+
+  const int numMesh = meshIDs.size();
+  const int* meshIDsPtr = meshIDs.cptrD();
+  int* originalPtr = original.ptrD();
+
+  thrust::for_each(thrust::device, outR.meshRelation_.triBary.begin(),
+                   outR.meshRelation_.triBary.end(),
+                   [=] __host__ __device__(BaryRef &b) {
```

Comment:
I need to get used to using these device lambdas, at least for simple functions like this. Would probably clean up the code a lot. When I started, I couldn't get these to work.